### PR TITLE
Discovery of Internet Gateways, RouteTables; ec2test improvements

### PR DESCRIPTION
**Pull request by dimitern**
Building on the previous PR, this one improves ec2test server and adds tests for some of its features (more in a follow-up). Added support to pre-configure ec2test servers with:
- SetInitialAttributes() got **renamed** to SetAccountAttributes(), and the behavior of treating "default-vpc" specially has been dropped in favor of AddDefaultVPCAndSubnets() below. In order not to break existing tests code using ec2test (if any), SetInitialAttributes() is an alias of SetAccountAttributes(), but to simulate a default VPC in ec2test such tests need to start calling AddDefaultVPCAndSubnets() instead.
- _VPCs_: AddVPC(), UpdateVPC(), RemoveVPC()
- AddDefaultVPCAndSubnets() - a convenient helper to add a default VPC, one default subnet per AZ (with MapPublicIPOnLaunch set), an Internet Gateway (IGW) connected to the VPC, a main route table for the VPC with routes for local VPC traffic and a default route to the IGW, finally calling SetAccountAttributes() to announce the default VPC and make it discoverable.
- _Subnets_: AddSubnet(), UpdateSubnet(), RemoveSubnet()
- _Internet Gateways_: AddInternetGateway(), UpdateInternetGateway(), RemoveInternetGateway()
- _Route Tables_: AddRouteTable(), UpdateRouteTable(), RemoveRouteTable()
- SetAvailabilityZones() now **removes** any subnets which have the zone in their AvailZone field.
- Reset() method added to ec2test, which takes a withoutZonesOrGroups flag - when `true` everything stored in the test server is reset, including counters, but the listener is kept. A value of `false` will add default AZ and security groups after resetting. 

In the ec2 package, added support for the DescribeInternetGateways and DescribeRouteTables AWS APIs, with more or less complete filtering support except for tags.

Finally, this PR includes a number of fixes around live tests to make them more robust and reliable.
Tested with different combinations of: `go test -check.v -test.timeout=1200m -amazon -race -coverprofile=cover -covermode=atomic`

### CODE REVIEW

File: ec2/ec2test/availability_zones.go [hunk 1]
```diff
@@ -17,15 +17,31 @@ import (
 	"gopkg.in/amz.v3/ec2"
 )
 
-// SetAvailabilityZones initializes the availability zones the test
+// SetAvailabilityZones replaces the availability zones the test
 // server is returning.
+//
+// NOTE: If zones does not contain one or more existing zones those
```
Comment (voidspace) [inline ec2/ec2test/availability_zones.go]: This comment is confusing to me. "If zones does not contain... those zones are removed". If it doesn't contain them, how can they be removed?

File: ec2/ec2test/availability_zones.go [hunk 2]
```diff
@@ -17,15 +17,31 @@ import (
...
+		if _, exists := oldZones[z.Name]; exists {
+			// Remove any subnets attached to this zone before
+			// removing it.
+			remainingSubnets := make(map[string]*subnet)
+			for _, sub := range srv.subnets {
+				if sub.AvailZone != z.Name {
+					remainingSubnets[sub.Id] = sub
+				}
+			}
+			srv.subnets = remainingSubnets
```
Comment (voidspace) [inline ec2/ec2test/availability_zones.go]: Shouldn't this be outside the loop? You replace it for every zone at the moment.

File: ec2/ec2test/availability_zones.go [hunk 3]
```diff
@@ -17,15 +17,31 @@ import (
...
+// associated with them!
 func (srv *Server) SetAvailabilityZones(zones []ec2.AvailabilityZoneInfo) {
 	srv.mu.Lock()
-	srv.zones = make([]availabilityZone, len(zones))
-	for i, z := range zones {
-		srv.zones[i] = availabilityZone{z}
+	defer srv.mu.Unlock()
+	oldZones := srv.zones
+	srv.zones = make(map[string]availabilityZone)
+	for _, z := range zones {
```
Comment (voidspace) [inline ec2/ec2test/availability_zones.go:31]: What this loop does is:

```
iterate over all the new zones removing all subnets associated with each new zone
```

Did you mean to iterate over oldZones instead?

(In the end, with the current loop, I believe that the only subnets left will be ones associated with zones from oldZones but _not_ in new zones. I think the effect you're after is exactly the opposite.)

File: ec2/ec2test/internet_gateways.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	if _, found := srv.internetGateways[igwId]; found {
+		delete(srv.internetGateways, igwId)
+		return nil
+	}
+	return fmt.Errorf("internet gateway %q not found", igwId)
+}
+
+type internetGateway struct {
```
Comment (voidspace) [inline ec2/ec2test/internet_gateways.go:82]: Why do we create a private type that exactly matches the public type - why not just use the public type?

File: ec2/ec2test/internet_gateways_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+	resp, err = s.ec2.InternetGateways([]string{added.Id}, nil)
+	c.Assert(err, IsNil)
+	c.Assert(resp, NotNil)
+	c.Assert(resp.InternetGateways, HasLen, 1)
+	c.Assert(resp.InternetGateways[0], DeepEquals, added)
+}
+
+func (s *S) TestUpdateInternetGateway(c *C) {
+	// AddDefaultVPCAndSubnets() creates a IGW as well.
```
Comment (voidspace) [inline ec2/ec2test/internet_gateways_test.go]: "an IGW"

Comment (voidspace) [inline ec2/ec2test/internet_gateways_test.go]: but you're not calling AddDefaultVPCAndSubnets are you? So either the comment is outdated or there is some code missing.

File: ec2/ec2test/availability_zones.go (hunk 3 above)
Comment (dimitern) [inline ec2/ec2test/availability_zones.go:31]: Good catch! You're correct, I'll change this.

File: ec2/ec2test/internet_gateways.go (hunk 4 above)
Comment (dimitern) [inline ec2/ec2test/internet_gateways.go:82]: internetGateway (and similar ec2test types) embed the public type to provide access to it, but the unexported type has extra (internal) methods (e.g. matchAttr called while filtering).

File: ec2/ec2test/internet_gateways_test.go (hunk 5 above)
Comment (dimitern) [inline ec2/ec2test/internet_gateways_test.go]: I'll fix the typo. AddDefaultVPCAndSubnets is called in SetUpTest of the S suite.

Comment (voidspace) [inline ec2/ec2test/internet_gateways_test.go]: Oh, I looked for a SetUp but didn't see one. I obviously didn't look hard enough! Thanks.

File: ec2/ec2test/route_tables_test.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+			State:                   "",
+			VPCPeeringConnectionId: "",
+		}, {
+			DestinationCIDRBlock: "0.0.0.0/0",
+			GatewayId:            "igw-0", // testing default VPC's IGW
+			State:                "active",
+		}},
+		PropagatingVGWIds: []string{"foo", "bar"},
+	}
+	added, err = s.srv.AddRouteTable(toAdd)
```
Comment (voidspace) [inline ec2/ec2test/route_tables_test.go:70]: Missing an assert about err.

File: ec2/ec2test/subnets_test.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	. "gopkg.in/check.v1"
+
+	"gopkg.in/amz.v3/ec2"
+)
+
+func (s *S) TestAddSubnet(c *C) {
+	subnet, err := s.srv.AddSubnet(ec2.Subnet{})
+	c.Assert(err, ErrorMatches, "empty VPCId field")
+	c.Assert(subnet, DeepEquals, ec2.Subnet{})
+	s.assertVPCsExist(c, true)
```
Comment (voidspace) [inline ec2/ec2test/subnets_test.go]: It's not clear why this assert is called here. If it's a precondition that should be true before AddSubnet and true after, then please add an additional call to assertVPCsExist before calling AddSubnet. If the call to AddSubnet _changes_ something about VPCs, please add a corresponding assert before AddSubnet.

File: ec2/internet_gateways_test.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	c.Assert(resp.RequestId, Equals, "59dbff89-35bd-4eac-99ed-be587EXAMPLE")
+	c.Assert(resp.InternetGateways, HasLen, 1)
+	igw := resp.InternetGateways[0]
+	c.Check(igw.Id, Equals, "igw-eaad4883EXAMPLE")
+	c.Check(igw.VPCId, Equals, "vpc-11ad4878")
+	c.Check(igw.AttachmentState, Equals, "available")
+	c.Check(igw.Tags, HasLen, 0)
+}
+
+// Internet Gateway tests that run either agsints the local test
```
Comment (voidspace) [inline ec2/internet_gateways_test.go]: "against" not agsints

**Comment by voidspace** [comment]
Other than a few minor points, LGTM

File: ec2/ec2test/route_tables_test.go (hunk 6 above)
Comment (dimitern) [inline ec2/ec2test/route_tables_test.go:70]: Thanks, will fix this.

File: ec2/ec2test/subnets_test.go (hunk 7 above)
Comment (dimitern) [inline ec2/ec2test/subnets_test.go]: Fair enough, will add a comment and move the assert (which just verifies AddDefaultVPCAndSubnets was called in SetUpTest) before AddSubnet.

File: ec2/internet_gateways_test.go (hunk 8 above)
Comment (dimitern) [inline ec2/internet_gateways_test.go]: Thanks, fill fix.

File: ec2/ec2test/network_interfaces.go [hunk 9]
```diff
@@ -404,6 +416,10 @@ func (srv *Server) createIFace(w http.ResponseWriter, req *http.Request, reqId s
 	return resp
 }
 
+func (srv *Server) dnsNameFromPrivateIP(privateIP string) string {
+	return fmt.Sprintf("ip-%s.ec2.internal", strings.Replace(privateIP, ".", "-", -1))
```
Comment (bz2) [inline ec2/ec2test/network_interfaces.go:420]: Using .internal may actually resolve to something when running tests, whereas .invalid is reserved, see rfc 2606. The tests should not be trying to do dns lookups on these faked names, but I like using .invalid or .test even so.

File: ec2/internet_gateways.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+//
+// See http://goo.gl/syjv2p for more details.
+type InternetGatewaysResp struct {
+	RequestId        string            `xml:"requestId"`
+	InternetGateways []InternetGateway `xml:"internetGatewaySet>item"`
+}
+
+// InternetGateways describes one or more Internet Gateways (IGWs).
+// Both parameters are optional, and if specified will limit the
+// returned IGWs to the matching ids or filterring rules.
```
Comment (bz2) [inline ec2/internet_gateways.go]: Tyop, 'filterring'.

**Comment by bz2** [comment]
LGTM, though without going in detail over all the new test code. As mentioned on irc, have some version compat concern with the ec2test changes but I'm not sure how widely that's used.

File: ec2/ec2test/network_interfaces.go (hunk 9 above)
Comment (dimitern) [inline ec2/ec2test/network_interfaces.go:420]: The only reason for this change was to make both the local-live and live tests pass consistently (AWS does not populate DNSName sometimes).

File: ec2/internet_gateways.go (hunk 10 above)
Comment (dimitern) [inline ec2/internet_gateways.go]: Fixed.

**Comment by dimitern** [comment]
Thanks for the reviews!
$$merge$$

**Pull request merged**